File: speedtestthread.cpp
#include <iostream>
#include <chrono>
#include <random>
#include <thread>

/*
speed test using multithreading
qty 2 prefilled array size [12k, 12k] multiplying into the first,
10 iterations(threaded)
consistent result: 875ms -- WOW!
*/

int const SIZE_X_ARRAY = 12000;
int const SIZE_Y_ARRAY = 12000;
int const NUM_ITERATIONS = 10;

// seed random from c++11 random
std::random_device rd;
std::mt19937 mt(rd());
std::uniform_real_distribution<double> dist(.0001, 2.000);


// run ea process as a thread
void process(double* aPtrs[], double* bPtrs[]){
    for(int y = 0; y < SIZE_Y_ARRAY; y++){
        for(int x = 0; x < SIZE_X_ARRAY; x++){
            aPtrs[y][x] *= bPtrs[y][x];
        }
    }
}

// might as well prefill as thread as well
void prefill(double* arrayPtrs[]){
    for(int y = 0; y < SIZE_Y_ARRAY; y++){
        for(int x = 0; x < SIZE_X_ARRAY; x++){
            //printf("%p\n", &(arrayPtrs[y][x])); // retrieve the address [y][x]
            arrayPtrs[y][x] = dist(mt);
        }
    }
}

int main(){
    // qty: 2, 2d arrays (pointer arrays of pointers)
    double* aPtrs[SIZE_Y_ARRAY];
    double* bPtrs[SIZE_Y_ARRAY];
    for(int y = 0; y < SIZE_Y_ARRAY; y++){
        aPtrs[y] = new double[SIZE_X_ARRAY];
        bPtrs[y] = new double[SIZE_X_ARRAY];
    }
    // array of threads (only for num iterations and processing threads)
    std::thread processThreads[NUM_ITERATIONS];
    
    
    // prefill threads (separate from above)
    std::thread aPre(prefill, aPtrs);
    std::thread bPre(prefill, bPtrs);
    // wait on both prefills
    aPre.join();
    bPre.join();
    std::cout << "done prefilling\n";


    // timer init
    using std::chrono::high_resolution_clock;
    using std::chrono::duration_cast;
    using std::chrono::duration;
    using std::chrono::milliseconds;
    auto t1 = high_resolution_clock::now();


    // process the work using threads
    for(int i = 0; i < NUM_ITERATIONS; i++){
        //process(aPtrs, bPtrs); // old non multi threaded
        processThreads[i] = std::thread(process, aPtrs, bPtrs);
    }
    // wait for all threads
    for(int i = 0; i < NUM_ITERATIONS; i++){
        processThreads[i].join(); // i cannot believe this works
    }
    
    
    //stop timer and print out
    auto t2 = high_resolution_clock::now();
    auto ms_int = duration_cast<milliseconds>(t2 - t1);
    std::cout << ms_int.count() << "ms\n";


    // debug test if it actually did the work
    /*
    for(int y = 0; y < SIZE_Y_ARRAY; y++){
        for(int x = 0; x < SIZE_X_ARRAY; x++){
            printf("%.8f\n", aPtrs[y][x]);
        }
    }
    */

}

File: speedtestthread.py
import numpy as np
import time
from multiprocessing.pool import ThreadPool as Pool

# speed test
# qty 2 prefilled array size [12k, 12k] multiplying into the first,
# multiproc version
# 20 iterations: 4120 4130 4280 4290 4300 4320 4520

Y_SIZE = 12000
X_SIZE = 12000
ITERATIONS = 20 # also number of pools in multiproc
pool = Pool(ITERATIONS)


def process(a_arr, b_arr):
    a_arr = np.multiply(a_arr, b_arr)
    return a_arr


if __name__ == "__main__":
    # create 2d array from random, it says it is a float64, and yet, it seems to capout at 8 decimals?
    a_arr = np.random.uniform(0, 2, (Y_SIZE, X_SIZE))
    b_arr = np.random.uniform(0, 2, (Y_SIZE, X_SIZE))
    #print(a_arr)
    #print(b_arr)
    print("done prefilling")
    # no actually it is float64, but it caps the values only when it prints it out
    #for y in range(Y_SIZE):
        #for x in range(X_SIZE):
            #print("hh", a_arr[y, x])


    # start timer and exec
    start = time.perf_counter()
    for iter in range(ITERATIONS):
        #a_arr = process(a_arr.view(), b_arr.view())
        # start async pools
        a_arr = pool.apply_async(process, (a_arr, b_arr))
    
    # waiting for all pools to finish
    pool.close()
    pool.join()
        
        
    # stop timer
    end = time.perf_counter()
    print("{:.3f}".format(end - start), "secs")
        
    #print(a_arr)